### PR TITLE
Drop `wheel` and `setuptools` from shared libs

### LINKED ISSUE #951: `pipx run wheel` / `install wheel` doesn't work
**Describe the bug**

`pipx run wheel` doesn't seem to be supported. I think it's because `wheel` is filtered out as part of the default environment.

~~A possible fix would be to add the entrypoint for pipx to wheel;~~ otherwise I think special handling would have to be included to see if 'wheel' is specifically asked for.

**How to reproduce**
`pipx run wheel` reports no executables found.

```console
$ docker run --rm -it python bash
# pip uninstall wheel
# pip install pipx
# pipx run wheel
'wheel' executable script not found in package 'wheel'.
Available executable scripts:

```

**Expected behavior**
It should get wheel.

I opened an example of the entrypoint in wheel, but I realized that `pipx install wheel` should work too, and it doesn't:

```console
# pipx install wheel

No apps associated with package wheel or its dependencies. If you are attempting to install a library, pipx should not be used. Consider using pip or a similar tool
instead.
```

So the fix needs to be here. Note that wheel isn't meant to be used as a library (at least yet), only as a CLI app (or a setuptools helper, of course). And the next version will be gaining a very useful CLI command (`tags`).

It also seems it's filtered from git?

```console
$ pipx run --spec git+https://github.com/henryiii/wheel.git@henryiii/fix/pipxrun wheel tag --help

Cannot determine package name from spec 'git+https://github.com/henryiii/wheel.git@henryiii/fix/pipxrun'. Check package spec for errors.
```

That might be related to the fact the version number is the same, and it doesn't reinstall the package, though.

### LINKED ISSUE #990: drop `setuptools` and `wheel` from shared libs
this would fix https://github.com/pypa/pipx/issues/951 and cause pip 23.1 to use the pep517 codepath for all packages

since 23.1 pip doesn't need wheel or setuptools anymore https://pip.pypa.io/en/stable/news/#v23-1

> When the wheel package is not installed, pip now uses the default build backend instead of setup.py install and setup.py develop for project without pyproject.toml. ([#8559](https://github.com/pypa/pip/issues/8559))

setuptools was also removed from CPython 3.12's ensurepip: https://docs.python.org/3.12/whatsnew/3.12.html#removed
> Remove the bundled setuptools wheel from [ensurepip](https://docs.python.org/3.12/library/ensurepip.html#module-ensurepip), and stop installing setuptools in environments created by [venv](https://docs.python.org/3.12/library/venv.html#module-venv).

### PR DESCRIPTION
**Pull request by chrysle**
And enable running and installing the remaining shared library `pip` using `pipx run`. I have transferred the hack from #970 per https://github.com/pypa/pipx/issues/951#issuecomment-1564513524 for use with `pip`.

 * Close #970 
 * Close #951 
 * Close #990

* [x] I have added an entry to `docs/changelog.md`

## Test plan

Tested by running
```
pipx run wheel
pipx run pip
pipx install pip
```

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 ## dev
 
+- Drop `setuptools` and `wheel` from the shared libraries
```
Comment (uranusjr) [inline CHANGELOG.md:3]: It may be a good idea to describe briefly what this means to end users.

Comment (chrysle) [reply CHANGELOG.md]: Sure, done. Have you already considered adopting `towncrier` for `pipx`? This would also simplify categorizing the changes.

**Review by gaborbernat — CHANGES_REQUESTED** [review]
CI is failing.

File: src/pipx/shared_libs.py [hunk 2]
```diff
@@ -107,8 +107,6 @@ def upgrade(
                         *_pip_args,
                         "--upgrade",
                         "pip",
```
Comment (webknjaz) [inline src/pipx/shared_libs.py:109]: This should set the minimum version of pip that doesn't need setuptools/wheel:
```suggestion
                         "pip >= 23.1",
```

File: src/pipx/commands/run.py [hunk 3]
```diff
@@ -227,6 +226,12 @@ def _download_and_run(
             package_or_url, python, pip_args=pip_args, verbose=verbose
         )
 
+    override_shared = False
+
+    if package_name == "pip":
+        override_shared = True
```
Comment (webknjaz) [inline src/pipx/commands/run.py:232]: Simplify?
```suggestion
    override_shared = package_name == "pip"
```

File: src/pipx/commands/install.py [hunk 4]
```diff
@@ -58,7 +58,12 @@ def install(
             return EXIT_CODE_INSTALL_VENV_EXISTS
 
     try:
-        venv.create_venv(venv_args, pip_args)
+        # Enable installing shared library `pip` with `pipx`
+        override_shared = False
+
+        if package_name == "pip":
+            override_shared = True
```
Comment (webknjaz) [inline src/pipx/commands/install.py:65]:
```suggestion
        override_shared = package_name == "pip"
```

**Review by webknjaz — COMMENTED** [review]
Got a few small suggestions..

**Comment by chrysle** [comment]
Thank you!

**Comment by gaborbernat** [comment]
We need to fix the CI first 😎

**Comment by dukecat0** [comment]
> We need to fix the CI first 😎

It should be fixed in #1117.

**Comment by dukecat0** [comment]
@chrysle Please fix linting.

**Comment by chrysle** [comment]
Done!

**Review by gaborbernat — APPROVED** [review]

**Comment by dukecat0** [comment]
Thanks!

**Comment by jaraco** [comment]
This change did not have the intended effect. Instead, it had the effect of leaving an old setuptools lingering in the shared libs, because that's what `venv` does:

```
root@1b9c9a77eeb4:/# python3.10 -m venv .venv
root@1b9c9a77eeb4:/# ls .venv/lib/python3.10/site-packages/
_distutils_hack  distutils-precedence.pth  pip  pip-22.0.2.dist-info  pkg_resources  setuptools  setuptools-59.6.0.dist-info
```

See #1539 for why that's worse than upgrading it.

**Pull request merged**